src/use_file.rs

```diff
@@ -38,36 +41,83 @@ pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
     Ok(())
 }
 
-cfg_if! {
-    if #[cfg(any(target_os = "android", target_os = "linux"))] {
-        fn init_file() -> Option<libc::c_int> {
-            // Poll /dev/random to make sure it is ok to read from /dev/urandom.
-            let mut pfd = libc::pollfd {
-                fd: unsafe { open_readonly("/dev/random\0")? },
-                events: libc::POLLIN,
-                revents: 0,
-            };
+// Returns the file descriptor for the device file used to retrieve random
+// numbers. The file will be opened exactly once. All successful calls will
+// return the same file descriptor. This file descriptor is never closed.
+fn get_rng_fd() -> Result<libc::c_int, Error> {
+    static FD: AtomicUsize = AtomicUsize::new(LazyUsize::UNINIT);
+    fn get_fd() -> Option<libc::c_int> {
+        match FD.load(Relaxed) {
+            LazyUsize::UNINIT => None,
+            val => Some(val as libc::c_int),
+        }
+    }
+
+    // Use double-checked locking to avoid acquiring the lock if possible.
+    if let Some(fd) = get_fd() {
+        return Ok(fd);
+    }
+
+    // SAFETY: Only a raw pointer is created from MUTEX to pass to libc.
+    static mut MUTEX: libc::pthread_mutex_t = libc::PTHREAD_MUTEX_INITIALIZER;
+    let r = unsafe { libc::pthread_mutex_lock(&mut MUTEX as *mut _) };
```

Comment:
I am not sure that `&mut MUTEX as *mut _` is formally sound. Here, we create a mutable reference to `MUTEX` *before* locking it, so, at runtime, we can get aliasing. The guaranteed ok solution is to put a mutex into a `Cell` / `UnsafeCell`, and store that in a non-mut static. `UnsafeCell` is not `Sync` though, so you'll need a custom wrapper struct to assert `Sync` (like I did in the playground link).

See also https://github.com/rust-lang/rust/issues/53639

Reply:
Isn't the purpose of mutexes to handle aliasing — i.e. that one needs to obtain a `*mut _` local to the thread (and aliasing the equivalent in any other thread) in order to use the mutex?

Reply:
Aha, the [`UnsafeCell`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html) docs clarify:

> The compiler makes optimizations based on the knowledge that &T is not mutably aliased or mutated, and that &mut T is unique.

>  The UnsafeCell API itself is technically very simple: it gives you a raw pointer *mut T to its contents. It is up to you as the abstraction designer to use that raw pointer correctly.

So it wrapping via `UnsafeCell` just allows us to skip the `&mut _` part.

Reply:
This should be fixed now. I added a `Mutex` abstraction that wraps an `UnsafeCell`. By adding `lock`/`unlock` methods to this `Mutex`, the code also ends up looking much cleaner.